core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackRequestValidator.java

```diff
@@ -69,5 +84,20 @@ private void validateTemplates(StackRequest stackRequest, ValidationResultBuilde
                 .reduce(ValidationResult::merge)
                 .ifPresent(resultBuilder::merge);
     }
+
+    private void validateSharedService(StackRequest stackRequest, ValidationResultBuilder validationBuilder) {
```

Comment:
The method is too complicated, there is a simplest way to solve the same problem:
```
private void validateSharedService(StackRequest stackRequest, ValidationResultBuilder validationBuilder) {
        if (stackRequest.getClusterToAttach() != null) {
            Stack stack = stackRepository.findOne(stackRequest.getClusterToAttach());
            if (stack != null && !AVAILABLE.equals(stack.getStatus())
                    || clusterService.retrieveClustersByStackId(stackRequest.getClusterToAttach()).stream().anyMatch(cluster -> !AVAILABLE.equals(cluster.getStatus())) {
                validationBuilder.error("Unable to attach to datalake cluster! Please check that the required datalake"
                    + " stack/cluster exists and all of it's nodes are in AVAILABLE status");
            }
        }
    }
```

Reply:
I would create to individual boolean variable to cover the complex expression of the If statement, that would be much easier to read and understand.

Reply:
i prefer `if (stack != null && !AVAILABLE.equals(stack.getStatus()) || anyClusterIsNotAvailable(stackRequest.getClusterToAttach())` because the sec expression evaluation is depends on the first.